Exclude Directories

mmattel opened the pull request:
Successor of PR https://github.com/owncloud/core/pull/16534

Before  mmattel/core
Now     owncloud/core

Shoud fix CI start

mmattel committed:
exclude directories from beeing processed

extended case: search at any place of the path given

adding case insensitiveness

added suggestions and improvements

removed unnecessary function parameter

new approach according the rethinking

fixed code and unit tests

update comments

improved code, added calls for trashbin copyFromStorage

mmattel committed:
exclude directories from beeing processed

extended case: search at any place of the path given

adding case insensitiveness

added suggestions and improvements

removed unnecessary function parameter

new approach according the rethinking

fixed code and unit tests

update comments

improved code, added calls for trashbin copyFromStorage

mmattel committed:
fix local storage

schiesbn commented on apps/files/ajax/newfile.php:
```diff
@@ -61,6 +61,15 @@
 	exit();
 }
 
+if (\OC\Files\Filesystem::isForbiddenFileOrDir($fileName)) {
```
shouldn't the hook catch this already?
https://github.com/owncloud/core/pull/19235/files#diff-73d71f74c9f6d7093f369006c0c60dfeR585

mmattel replied:
No it does not.
And we need this query here to display on the UI the text.

schiesbn replied:
What's about "upload then? There you also get a generic error message only?
Maybe it would be a good idea to let the hook throw an exception with the corresponding error message if it is forbidden and then make sure that we display the error message from the exception instead of a generic one when we catch the exception. You could use the HintException to provide a translated message. We do something similar with encryption errors.

It is probably more failure save to add if statements all over the place and it is more generic since other errors can be handled the same way.

schiesbn commented on apps/files/ajax/newfolder.php:
```diff
@@ -53,6 +53,15 @@
 	return;
 }
 
+if (\OC\Files\Filesystem::isForbiddenFileOrDir($folderName)) {
```
shouldn't the hook catch this already?
https://github.com/owncloud/core/pull/19235/files#diff-73d71f74c9f6d7093f369006c0c60dfeR585

mmattel replied:
No it does not.
And we need this query here to display on the UI the text.

oparoz commented:
@mmattel @PVince81 @rperezb In order to test this properly, could we have some sort of acceptance test list with checkboxes? Would make it easier to know what's been tested and which corner cases we may have forgotten about.

mmattel commented:
**Testscenario_1:**
Focus: try to create a excluded directory name
Preperation: the name does NOT exist in the filesystem

Storage | Task (t) | Browser | Webdav | Creatable | Result (r) | Comment
---------- | ---------- | ------------ | ----------- | ------------- | -------- | ------------ |
local     |  D	    | x	    	     |		     |	yes		| 	    |   |
local     |  E	    | x    	     |		     |	no		| 	    | (1) |
local     |  D	    |		     |	x	     |	yes		| 	    |  |
local     |  E	    |		     |	x	     |	no		| 	    | (2) |
SMB     |  D	    | x		     |		     | yes		| 	    |  |
SMB     |  E	    | x		     |		     | no		| 	    |	(1) |
SMB     |  D	    |		     |	x	     |	yes		| 	    |  | 
SMB     |  E	    |		     |	x	     |	no		| 	    |	(2) |


* t ...  define a directory name to be excluded like ``fgh``
	- E ... enable it config.php
	- D ... disable it in config.php (I just renamed it to ``.fgh`` and vice versa)
* r ... y (yes), n (no), -- (n/a)

(1) ... A browser popup should appear that this name is excluded and can´t be used
(2) ... WebDav reports 'Forbidden' 

- [ ] Testscenario_1 sucessfully finished

**Testscenario_2:**
Focus: are the directories properly shown.
Preperation: have a directory already created with an excluded name which will then be checked against. This means, that the ecluded directory name already exists in the filesystem.

Storage | Task (t) | Browser | Webdav | Is_Visible | Result (r) | Comment
---------- | ---------- | ------------ | ----------- | ------------- | -------- | ------------ |
local     |  D	    | x	    	     |		     |	yes		| 	    |   |
local     |  E	    | x    	     |		     |	no		| 	    |   |
local     |  D	    |		     |	x	     |	yes		| 	    |   |
local     |  E	    |		     |	x	     |	no		| 	    |   |
SMB     |  D	    | x		     |		     | yes		| 	    |   |
SMB     |  E	    | x		     |		     | no		| 	    |   |
SMB     |  D	    |		     |	x	     |	yes		| 	    |   | 
SMB     |  E	    |		     |	x	     |	no		| 	    |   |


* t ...  define a directory name to be excluded like ``fgh``
	- E ... enable it config.php
	- D ... disable it in config.php (I just renamed it to ``.fgh`` and vice versa)
* r ... y (yes), n (no), -- (n/a)


- [ ] Testscenario_2 sucessfully finished

**Testscenario_3:**
Focus: try to create an excluded directory name where the name DOES already exist in the filesystem
Preperation: like test 2

Storage | Task (t) | Browser | Webdav | Is_Creatable | Result (r) | Comment
---------- | ---------- | ------------ | ----------- | ------------- | -------- | ------------ |
local     |  D	    | x	    	     |		     |	no		| 	    | same as Test2 |
local     |  E	    | x    	     |		     |	no		| 	    | same as Test2 |
local     |  D	    |		     |	x	     |	no		| 	    | (1) |
local     |  E	    |		     |	x	     |	no		| 	    | (1) |
SMB     |  D	    | x		     |		     |  no		| 	    | same as Test2 |
SMB     |  E	    | x		     |		     |  no		| 	    | same as Test2 |
SMB     |  D	    |		     |	x	     |	no		| 	    | (1) | 
SMB     |  E	    |		     |	x	     |	no		| 	    | (1) |


* t ...  define a directory name to be excluded like ``fgh``
	- E ... enable it config.php
	- D ... disable it in config.php (I just renamed it to ``.fgh`` and vice versa)
* r ... y (yes), n (no), -- (n/a)

(1) ... A Fatal Error Message is created, the WebDav client also shows an error. But this is according to my tests not related to this PR. For details pls see https://github.com/owncloud/core/issues/19101 (WebDav: create directory with an already existing directory name -> Fatal error) 


- [ ] Testscenario_3 sucessfully finished

Note : I wanted to use checkboxes in the table, but this did not work

oparoz commented:
Excellent, thank you :)

MorrisJobke commented:
@DeepDiver1975 @oparoz I don't think we want this in 8.2. We are already in feature freeze.

oparoz commented:
I don't have a say in this ;)

MorrisJobke commented:
@DeepDiver1975 @cmonteroluque enhancement -> 9.0

cmonteroluque commented:
@MorrisJobke ok

DeepDiver1975 commented:
rebased

PVince81 commented:
@icewind1991 can you please comment on this PR ?

DeepDiver1975 commented:
needs rebase

RealRancor commented:
Please also ping in https://github.com/owncloud/documentation/pull/1275 if this PR gets merged.

icewind1991 commented:
The main problem i have left with this is the seperate in the config between "blacklisted files" and "excluded directories", I see no reason why they can't be the came config entry

Also `isForbiddenFileOrDir` looks like it could be a lot easier

Something like

```php
$config = \OC::$server->getSystemConfig();
$excluded = $config->getValue('blacklisted_files', ['.htaccess']);
$path  = str_replace('\\','/', $path);
$paths  = explode('/', $path);
return (bool) array_intersect($parts, $excluded);
```

should be enough

not sure about the making the blacklist case insensitive

mmattel commented:
> 	* Blacklist ... files that may harm the owncloud environment like a foreign .htaccess file
>	* Excluded  ... directories that are excluded from beeing further processed, like snapshot directories

There is a big difference between those two parameters and the meaning behind.

**Excluded Directories** mean, that the storage or filesystem mounted may contain folders having a part in the path name which you want to exclude from scanning (therefore the name excluded) and from an oC point of view disallow creation or viewing or renaming to with any access method. When  scanning occured and the config parameter was not set but the storage or filesystem contained such directories, after setting the parameter, the must be excluded from further processing. (pls also read #16534)

**Blacklisted Files** mean, they are files that can be present but would impact owncloud and therefore be blacklisted.

I highly recommend keeping those two parameters seperated, because this gives admins the proper meaning what they want to do.

Regarding ``isForbiddenFileOrDir``, any improvment is positive. Excluded Directories must be case insensitive and should be trimmed. The only thing to consider is, that we do not miss any possible case. The test highlights all possible cases I have identified:

````
	public function isExcludedData() {
		return array(
			array('.snapshot', true),
			array('.snapshot/', true),
			array('.snapshot\\', true),
			array('/etc/foo/bar/foo.txt', false),
			array('/.snapshot/etc/foo/bar/foo.txt', true),
			array('/.snapShot/etc/foo/bar/foo.txt', true),
			array('\.snapshot\etc\foo/bar\foo.txt', true),
			array('/etc/foo/.snapshot', true),
			array('/etc/foo/.snapshot/bar', true),
		);
````

Note: due to time constraints I have very limited possibilities to continue on that, but would be happy if that idea would get implemented. There is a sound need for and many request made!
Please see also the test scenarious above I created.

DeepDiver1975 commented:
@mmattel rebase is required please

@icewind1991 accept or not? THX

PVince81 commented:
@DeepDiver1975 this has been rebased many many times by @mmattel already and he stated that he doesn't have time to continue this.

So we need to make a decision about this PR and take over.

PVince81 commented:
@MTRichards I have added this to the planning list https://github.com/owncloud/core/issues/24684#issuecomment-222628562

Hopefully we can finally get this merged in the next release

PVince81 commented on apps/dav/lib/connector/sabre/directory.php:
```diff
@@ -85,7 +85,19 @@ class Directory extends \OCA\DAV\Connector\Sabre\Node
+			# the check here is necessary, because createFile uses put covered in sabre/file.php 
+			# and not touch covered in files/view.php
+			if (\OC\Files\Filesystem::isForbiddenFileOrDir($name)) {
+				throw new \Sabre\DAV\Exception\Forbidden();
```
@mmattel any reason the check is done twice within the same method, outside and inside the try ?
Looks like a potential rebase mistake.

mmattel replied:
You are right, not necessary to have it twice. Outside should be fine. Thanks for picking this up. 
One thing I remember was the challenge to find a place where all DAV filesystem calls are passing thru, so this query would only be present once, but with the test I made this was not the case or at least I failed finding it...
The point is to cover read, write, rename, (move) and check if a isForbiddenFileOrDir is requested.
Easier for standard filesystem access, but seems to be more complex for DAV.

I would be very happy if you take over and finalize this PR :smiley: (I am really sorry for but I am lacking the time for)

PVince81 commented:
I'm wondering whether this would be easier using a storage wrapper and throw the exception from there for any file operation. But then it would make the exclusion relative to the storage roots instead of the user's home.

Anyway, I'm willing to take over this at some point.

mmattel commented:
> But then it would make the exclusion relative to the storage roots instead of the user's home.

Hope not blaming myself, but does this make a difference?
I come from another point of view: whenever there is a filesystem request coming from any source, check for a forbidden pattern in any place in the given string and reject processing it if true. Any working method for achieving this is valid.

PVince81 commented:
> but does this make a difference?

It only makes a difference on how the admin wants to configure it.

Anyway, it was mostly an additional thought I wanted to document here. I think your approach is fine.

PVince81 commented:
Rebased, not tested yet (files app still seems to work at the moment)

DeepDiver1975 commented:
unit tests are badly falling apart - 89 failed tests
https://jenkins.owncloud.org/job/owncloud-core/job/core/job/PR-19235/1/testReport/

mmattel commented on lib/private/Files/View.php:
```diff
@@ -1408,6 +1408,9 @@ public function getDirectoryContent($directory, $mimetype_filter = '') {
 			 */
+			$files = array_filter(function(ICacheEntry $content) {
+				return (!\OC\Files\Filesystem::isForbiddenFileOrDir($content['path']));
+			});
```
Should´nt that be (without testing):
```
$files = array_filter($content, function(ICacheEntry $c) {
				return (!\OC\Files\Filesystem::isForbiddenFileOrDir($c['path']));
			});
```

PVince81 replied:
yes, fixed

Commit:
Fix rebase issue in view

PVince81 commented:
Did a quick test with the old Webdav and new Webdav endpoints and an excluded folder called "snapshots".

Seems to work so far: listing, deleting or overwriting the folder is denied.

mmattel commented:
Great !
I have written test cases, pls see above. If they pass then this PR should be ready. Sorry if I cant do, businesstrip...

PVince81 commented:
> businesstrip...

No time to kill on the plane ? (if applicable) :wink: 

Am thinking of maybe writing integration tests to automatically test this.
The browser test cases can be removed as the web UI uses Webdav already since 9.0 (except for uploads)

mmattel commented:
@PVince81 any news? would really appreciate to get that in

PVince81 commented:
Tested with many file operations on local storage and external storage and also subdirectories.
Sync client works too.
Any attempt to do any file operation on the excluded folder results in 403 Forbidden.

Exclusion works fine :+1:

PVince81 commented:
- added to 9.2 feature list: https://github.com/owncloud/core/wiki/ownCloud-9.2-Features
- added doc ticket: https://github.com/owncloud/documentation/issues/2603

PVince81 commented:
@mmattel thanks a lot ! And sorry that it took so long to get in.

mmattel commented:
:tada: :smiley:

Kawohl commented:
@Pvince is this safe to patch into current 9.1 ?

PVince81 commented:
@Kawohl no, this is a huge change and I wouldn't recommend patching. This would need careful testing.

Kawohl commented:
thanks!
